privacyidea/lib/tokens/foureyestoken.py

```diff
@@ -277,3 +330,181 @@ def authenticate(self, passw, user=None, options=None):
+                if challengeobject.is_valid():
+                    # challenge is still valid
+                    used_tokens = pickle.loads(challengeobject_list[0].data or pickle.dumps({}))
+                    remaining_realms = self._get_remaining_realms(used_tokens)
+                    r_success = self._authenticate_remaining_realms(passw, remaining_realms, used_tokens, options)
+
+                    if r_success:
+                        challengeobject.set_otp_status(True)
+                    if not r_success:
+                        # increase the received_count
+                        challengeobject.set_otp_status()
+
+        self.challenge_janitor()
+        return r_success
+
+    def create_challenge(self, transactionid=None, options=None):
+        """
+        This method creates a challenge, which is submitted to the user.
+        The submitted challenge will be preserved in the challenge
+        database.
+
+        The challenge is a randomly selected question of the available
+        questions for this token.
+
+        If no transaction id is given, the system will create a transaction
+        id and return it, so that the response can refer to this transaction.
+
+        :param transactionid: the id of this challenge
+        :param options: the request context parameters / data
+        :type options: dict
+        :return: tuple of (bool, message, transactionid, attributes)
+        :rtype: tuple
+
+        The return tuple builds up like this:
+        ``bool`` if submit was successful;
+        ``message`` which is displayed in the JSON response;
+        additional ``attributes``, which are displayed in the JSON response.
+        """
+        options = options or {}
+        attributes = {}
+        if type(options.get("data")) == dict:
+            # In the special first chal-resp case we do not have pickled data, yet. So we need to convert
+            options["data"] = pickle.dumps(options.get("data"))
+        used_tokens = pickle.loads(options.get("data", pickle.dumps({})))
+        remaining_realms = self._get_remaining_realms(used_tokens)
+        if remaining_realms:
+            message = "Please authenticate with another token from " \
+                      "either realm: {0!s}.".format(", ".join(remaining_realms))
+
+        validity = int(get_from_config('DefaultChallengeValidityTime', 120))
+        tokentype = self.get_tokentype().lower()
+        # Maybe there is a QUESTIONChallengeValidityTime...
+        lookup_for = tokentype.capitalize() + 'ChallengeValidityTime'
+        validity = int(get_from_config(lookup_for, validity))
+
+        # Create the challenge in the database
+        db_challenge = Challenge(self.token.serial,
+                                 transaction_id=transactionid,
+                                 data=options.get("data"),
+                                 session=options.get("session"),
+                                 challenge=message,
```

Comment:
``message`` might not be initialized here.

Reply:
done in c0711790